### PR TITLE
Anasazi: Added Randomized Eigensolver

### PR DESCRIPTION
**Pull request by Heatherms27**
@trilinos/anasazi

Requested Reviewers:
Jennifer Loe
Erik Boman

## Motivation

We added into Anasazi a new solver that creates a block of orthonormal random initial starting vectors before performing subspace iteration by applying a matrix A to the block a specified number of times. The user can also specify how often to orthonormalize the basis + check the residuals for convergence. 
The purpose of this solver is to run a cheap eigensolver that sacrifices accuracy for speed. It is currently being tested in conjunction to spectral graph partitioning in Sphynx under the Zoltan2 package.

## Testing

Directly testing in:
/Trilinos/packages/anasazi/tpetra/test/Randomized/rand_driver.cpp
with the executable
/Trilinos/build/packages/anasazi/tpetra/test/Randomized/Anasazi_Randomized_Solver_Driver.exe

The randomized solver has almost been added as an option into:
/Trilinos/packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp
with the executable
Trilinos/build/packages/zoltan2/test/sphynx/Zoltan2_Sphynx_Research_Driver.exe

## Additional Information
Have also added more test files which are identical to pre-existing "cxx_main_doublefloat.cpp" tests , but allow for Matrix Market files as input.
/anasazi/tpetra/test/LOBPCG/cxx_main_test.cpp -> Anasazi_Tpetra_LOBPCG_test.exe
/anasazi/tpetra/test/BlockKrylovSchur/cxx_main_test.cpp -> Anasazi_Tpetra_BlockKrylovSchur_test.exe
/anasazi/tpetra/test/BlockDavidson/cxx_main_test.cpp -> Anasazi_Tpetra_BlockDavidson_test.exe

### CODE REVIEW

**Comment by jennloe** [comment]
Added pre-test inspection so the autotester can run.

File: packages/anasazi/tpetra/test/BlockKrylovSchur/cxx_main_test.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+// I/O for Matrix Market
+#include <MatrixMarket_Tpetra.hpp>
+
+using Tpetra::CrsMatrix;
+using Tpetra::Map;
+using std::vector;
+
+int main(int argc, char *argv[])
+{
+#ifndef HAVE_TPETRA_COMPLEX_DOUBLE
```
Comment (hkthorn) [inline packages/anasazi/tpetra/test/BlockKrylovSchur/cxx_main_test.cpp:73]: It looks like this test is for double precision Matrix Market files, so why does HAVE_TPETRA_COMPLEX_DOUBLE need to be defined?

File: packages/anasazi/tpetra/test/Randomized/Dummy
Comment (hkthorn) [inline packages/anasazi/tpetra/test/Randomized/Dummy:1]: Don't think that these files are needed anymore for copying test matrices into the build tree.

File: packages/anasazi/src/AnasaziRandomizedSolMgr.hpp [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+
+#include "Teuchos_TimeMonitor.hpp"
+#include "Teuchos_FancyOStream.hpp"
+#include "Teuchos_LAPACK.hpp"
+
+/*! \class Anasazi::RandomizedSolMgr
+  \brief The Anasazi::RandomizedSolMgr approximates largest eigenvalues/eigenvectors
+  by performing a simple Rayleigh-Ritz projection over a random block of vectors. 
+
+  The algorithms is well-known. See \\TODO citations.
```
Comment (jennloe) [inline packages/anasazi/src/AnasaziRandomizedSolMgr.hpp:77]: Add citations here.

**Review by jennloe — CHANGES_REQUESTED** [review]
We've also talked about removing some unneeded files.

**Comment by Heatherms27** [comment]
All issues Heidi and Jennifer mentioned should be resolved now. I want to note that the nonhermitian matrix (mhd1280a.cua) used as a default in the file anasazi/tpetra/test/BlockKrylovSchur/cxx_main_complex.cpp caused the autotester script to fail, as it did not converge. For now, I've just set the default matrix to be "mhd1280b.cua" to get the autotester to not return any errors.

The warnings CDash returned are all:
1. "variable not used" warnings from files I have never touched.
2. "warning: libgfortran.so.3, needed by /usr/lib64/liblapack.so, may conflict with libgfortran.so.5"
@jennloe @hkthorn

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 3]
```diff
@@ -54,27 +9,37 @@
 
 #include "Teuchos_TimeMonitor.hpp" 
 #include "Teuchos_StackedTimer.hpp"
-#include "readMatrixFromBinaryFile.hpp"
+#include "ReadMatrixFromFile.hpp"
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:12]: This needs to be changed back to the prior.  "readMatrixFromBinaryFile.hpp" does exist in the directory. Please remove the "ReadMatrixFromFile.hpp"

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 4]
```diff
@@ -1,48 +1,3 @@
...
-//
-// Questions? Contact Seher Acer        (sacer@sandia.gov)
-//                    Erik Boman        (egboman@sandia.gov)
-//                    Siva Rajamanickam (srajama@sandia.gov)
-//                    Jennifer Loe      (jloe@sandia.gov)
-//
-// ***********************************************************************
-//
-// @HEADER
 #include "Teuchos_CommandLineProcessor.hpp"
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:1]: The above copyright statement has to be added back. It cannot be removed.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 5]
```diff
@@ -54,27 +9,37 @@
...
 #include <Galeri_XpetraProblemFactory.hpp>
 #include <Galeri_XpetraParameters.hpp>
 #include <MatrixMarket_Tpetra.hpp>
 
-/////////////////////////////////////////////////////////////////////////////
-// This is a driver with many available options that can be used to test
-// a variety of features of Sphynx.
-// Note: This is research code. We do not guarantee it is without bugs. 
-/////////////////////////////////////////////////////////////////////////////
-
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:69]: Please add back this message.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 6]
```diff
@@ -54,27 +9,37 @@
...
+#include "AnasaziBasicEigenproblem.hpp"
+#include "AnasaziSimpleLOBPCGSolMgr.hpp"
+#include "AnasaziBasicOutputManager.hpp"
+#include "AnasaziEpetraAdapter.hpp"
+#include "Epetra_CrsMatrix.h"
+//#include "Teuchos_CommandLineProcessor.hpp"
+#include "Teuchos_Assert.hpp"
+#include "AnasaziTypes.hpp"
+*/
+
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:33]: Please remove any code that isn't needed.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 7]
```diff
@@ -338,136 +217,150 @@ int main(int narg, char *arg[])
...
+      if(me == 0 && verbosity > 1) std::cout << "Used Seher's reader for Largest Comp." << std::endl;
     }
     else if(std::equal(mtx.rbegin(), mtx.rend(), matrix_file.rbegin())) {
       typedef Tpetra::MatrixMarket::Reader<crs_matrix_type> reader_type;
       reader_type r;
       tmatrix = r.readSparseFile(matrix_file, pComm);
-      if (me==0){
-        std::cout << "Used standard Matrix Market reader." << std::endl;
-      }
+      if(me == 0 && verbosity > 1) std::cout << "Used standard reader." << std::endl;
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:327]: Please add "Matrix Market" in the message for clarity.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 8]
```diff
@@ -338,136 +217,150 @@ int main(int narg, char *arg[])
...
     }
-
-    Teuchos::RCP<const map_type> map = tmatrix->getMap();
-
-    Teuchos::RCP<mv_type> V;
-    if (vector_file !=""){
-      V = Tpetra::MatrixMarket::Reader<mv_type >::readDenseFile(vector_file,pComm,map);
-      if(me == 0){
-        std::cout << "Done with reading user-provided eigenvectors." << std::endl;
-      }
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:469]: Can you put this back? We used it for testing before.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 9]
```diff
@@ -338,136 +217,150 @@ int main(int narg, char *arg[])
...
+      V->randomize();
+    // std::cout << "Randomized some pretend evects." << std::endl;
     }
+    else{
+    V = Tpetra::MatrixMarket::Reader<MultiVector >::readDenseFile(vector_file,pComm,map);
+    Teuchos::RCP<const Tpetra::Map<> > vector_map = V->getMap();
+    if(me == 0)
+    std::cout << "Done with reading/creating the eigenvector." << std::endl;
+    }*/
+    // TODO Insert MultiVector Reader
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:362]: This you can remove.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 10]
```diff
@@ -516,50 +410,43 @@ int main(int narg, char *arg[])
...
+          Teuchos::TimeMonitor t3(*Teuchos::TimeMonitor::getNewTimer("Partitioning::Solve"));
           if (vector_file ==""){
             if(me == 0)
               std::cout << eigensolve << "will be used to solve the partitioning problem." << std::endl;
             problem->solve();
           }
           else{
-            std::cout << "Problem to be partitioned with user-provided eigenvectors." << std::endl;
-            problem->setUserEigenvectors(V);
-            problem->solve();
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:548]: Does this work if we put it back?

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 11]
```diff
@@ -516,50 +410,43 @@ int main(int narg, char *arg[])
...
       if (prec != "") sphynxParams->set("sphynx_preconditioner_type", prec);
       if (tol != -1) sphynxParams->set("sphynx_tolerance", tol);
-      
+
       using problem_type = Zoltan2::SphynxProblem<adapter_type>; //We found sphynx
       Teuchos::RCP<problem_type> problem;
       pComm->barrier();
       {
-        Teuchos::TimeMonitor t1b(*Teuchos::TimeMonitor::getNewTimer("Partitioning::All"));
+        Teuchos::TimeMonitor t(*Teuchos::TimeMonitor::getNewTimer("Partitioning::All"));
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:427]: Wasn't there a timer `t` defined earlier?

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp [hunk 12]
```diff
@@ -492,14 +386,14 @@ int main(int narg, char *arg[])
...
+      using problem_type = Zoltan2::PartitioningProblem<adapter_type>;
       Teuchos::RCP<problem_type> problem;
       pComm->barrier();
       {
-        Teuchos::TimeMonitor t1(*Teuchos::TimeMonitor::getNewTimer("Partitioning::All"));
+        Teuchos::TimeMonitor t(*Teuchos::TimeMonitor::getNewTimer("Partitioning::All"));
         {
           Teuchos::TimeMonitor t2(*Teuchos::TimeMonitor::getNewTimer("Partitioning::Problem"));
-          problem = Teuchos::rcp(new problem_type(adapter.getRawPtr(), params.getRawPtr(), sphynxParams, Tpetra::getDefaultComm()));
+          problem = Teuchos::rcp(new problem_type(adapter.getRawPtr(), params.getRawPtr(), Tpetra::getDefaultComm()));
```
Comment (jennloe) [inline packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:396]: So where are the sphynxParams getting passed in?

File: packages/zoltan2/sphynx/src/Zoltan2_SphynxProblem.hpp [hunk 13]
```diff
@@ -231,59 +232,87 @@ static void setSphynxValidatorsInList(
...
+        const std::string &algorithm, const std::string &defString,
+        const std::string &model, Environment &env, bool &removeSelfEdges,
+        bool &isGraphType, bool &needConsecutiveGlobalIds)
+    {
+      this->algName_ = std::string("sphynx");
+    }
+
+  /*template <typename Adapter>
+    void SphynxProblem<Adapter>::solve(Teuchos::RCP<mvector_t> &userEigenVects)
+    {
```
Comment (jennloe) [inline packages/zoltan2/sphynx/src/Zoltan2_SphynxProblem.hpp:245]: We need to either make this work or remove it.

File: packages/anasazi/tpetra/test/BlockKrylovSchur/cxx_main_complex.cpp [hunk 14]
```diff
@@ -125,13 +127,15 @@ int main(int argc, char *argv[])
   }
   if (debug) verbose = true;
   if (filename == "") {
-    // get default based on herm
+    filename = "mhd1280b.cua";
+/*    // get default based on herm
```
Comment (jennloe) [inline packages/anasazi/tpetra/test/BlockKrylovSchur/cxx_main_complex.cpp:131]: Remove commented code

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp (hunk 11 above)
Comment (Heatherms27) [reply packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:427]: Yes, but they are in an if/else statement, so they won't conflict with each other.

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp (hunk 12 above)
Comment (Heatherms27) [reply packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:396]: This isn't using Sphynx, this particular call is using parmetis or pulp. Do I still add in the parameters in that case?

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp (hunk 10 above)
Comment (Heatherms27) [reply packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:548]: No, but I'm trying to figure out why

File: packages/zoltan2/sphynx/src/Zoltan2_SphynxProblem.hpp (hunk 13 above)
Comment (Heatherms27) [reply packages/zoltan2/sphynx/src/Zoltan2_SphynxProblem.hpp:245]: I think it's best to remove it. I tried to debug the errors I got when I uncommented this code, but with no success

File: packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp (hunk 10 above)
Comment (Heatherms27) [reply packages/zoltan2/test/sphynx/Sphynx_Research_Driver.cpp:548]: Deleted the option to pass in user-specified eigenvectors.

**Comment by jennloe** [comment]
@hkthorn I am now happy with the changes in zoltan2/sphynx. We can merge this if you are happy with the Anasazi changes.  (I think Heather's main changes to the Anasazi tests were to add an option to read in Matrix Market files. Unfortunately, the git diff also caught a lot of reformatting.)

**Comment by hkthorn** [comment]
@jennloe I'm good with the changes now.  I think the commits should be squashed.  :)

**Review by jennloe — APPROVED** [review]

**Comment by sebrowne** [comment]
I recommend rebasing this branch given the 'develop' history change.

**Review by jennloe — APPROVED** [review]

**Comment by cgcgcg** [comment]
Please squash the changes.

**Comment by sebrowne** [comment]
> Please squash the changes.

Agreed, looking at the development history of this branch, that is likely the easiest approach.  That should resolve the LFS check issue.

**Comment by cgcgcg** [comment]
Please squash.

**Comment by jennloe** [comment]
@Heatherms27 something messed up here... tests are failing now that had been passing before.

**Review by egboman — APPROVED** [review]
Looks good. Thanks!

**Comment by jennloe** [comment]
Closing this PR. Merging in a cleaned-up version of this in PR #12414

**Pull request closed without merging**